sw/device/silicon_creator/lib/manifest.h

```diff
@@ -0,0 +1,210 @@
+   * obtained using `manifest_signed_region_get`.
+   */
+  sigverify_rsa_buffer_t signature;
+  /**
+   * Image length in bytes.
+   */
+  uint32_t image_length;
+  /**
+   * FIXME: Replace with max_version, min_version.
+   */
+  uint32_t image_version;
+  /**
+   * Image timestamp.
+   */
+  uint64_t image_timestamp;
+  /**
+   * Exponent of the signer's RSA public key.
+   */
+  uint32_t exponent;
+  /**
+   * FIXME: remove this field.
+   */
+  uint32_t reserved1;
+  /**
+   * FIXME: Replace these with binding_tag and max_key_version.
+   */
+  uint32_t usage_constraints[8];
+  uint32_t lockdown_info[4];
+  /**
+   * Modulus of the signer's RSA public key.
+   */
+  sigverify_rsa_buffer_t modulus;
+  /**
+   * Extension fields.
+   * FIXME: Remove these until we have a clear use-case.
+   */
+  uint32_t extension0_offset;
+  uint32_t extension0_checksum;
+  uint32_t extension1_offset;
+  uint32_t extension1_checksum;
+  uint32_t extension2_offset;
+  uint32_t extension2_checksum;
+  uint32_t extension3_offset;
+  uint32_t extension3_checksum;
+} manifest_t;
+
+static_assert(offsetof(manifest_t, identifier) == 0, "");
+static_assert(offsetof(manifest_t, reserved0) == 4, "");
+static_assert(offsetof(manifest_t, signature) == 8, "");
+static_assert(offsetof(manifest_t, image_length) == 392, "");
+static_assert(offsetof(manifest_t, image_version) == 396, "");
+static_assert(offsetof(manifest_t, image_timestamp) == 400, "");
+static_assert(offsetof(manifest_t, exponent) == 408, "");
+static_assert(offsetof(manifest_t, reserved1) == 412, "");
+static_assert(offsetof(manifest_t, usage_constraints) == 416u, "");
+static_assert(offsetof(manifest_t, lockdown_info) == 448u, "");
+static_assert(offsetof(manifest_t, modulus) == 464u, "");
+static_assert(offsetof(manifest_t, extension0_offset) == 848u, "");
+static_assert(offsetof(manifest_t, extension0_checksum) == 852u, "");
+static_assert(offsetof(manifest_t, extension1_offset) == 856u, "");
+static_assert(offsetof(manifest_t, extension1_checksum) == 860u, "");
+static_assert(offsetof(manifest_t, extension2_offset) == 864u, "");
+static_assert(offsetof(manifest_t, extension2_checksum) == 868u, "");
+static_assert(offsetof(manifest_t, extension3_offset) == 872u, "");
+static_assert(offsetof(manifest_t, extension3_checksum) == 876u, "");
+static_assert(sizeof(manifest_t) == 880, "");
+
+/**
+ * Helper macro for defining the manifest of a boot stage stored in flash.
+ *
+ * This definition places the manifest struct in the `.manifest` section, which
+ * is the first section of boot stages stored in flash, e.g. see
+ * sw/device/silicon_creator/rom_exts/rom_ext_common.ld. The `used` attribute
+ * ensures that the compiler emits the definition even if it is not used.
+ *
+ * Boot stages should not need to access their own manifests but since the
+ * manifest is updated using the signer tool, the manifest struct is
+ * volatile-qualified in case this assumption changes in the future.
```

Comment:
I'm skeptical about the volatile qualification.  Are you concerned that the compiler would inline the zero-initialized field values as created by the `const` declaration with no initializer-list?

Reply:
I think the correct way to do this is to actually do something like

```c
extern const volatile manifest_t name;
```

and then define the symbol in the linker script like so:

```
SECTIONS {
  .manifest : {
    kManifest = .;
    . += 880;
  } > flash;
}
```

I don't think we need this to be volatile? I don't think the compiler, even in LTO mode, can see through linker-defined symbols. I'll check what other LLVM people think... @asb  got any opinions?

Reply:
Sure, but defining the struct using the section attribute has the advantage that we don't have to specify the size in the linker script. Can you think of any downsides to using the section attribute (maybe besides using an attribute)?

Reply:
Avoiding specifying the size in the linker script (and to be honest, avoiding having to modify linker scripts at all) seems like an advantage for Alphan's current proposal.

I think `volatile` could be safely left off @mcy's suggestion for current Clang+LLVM, but there was previously talk about LTO+linker scripts (see [RFC](https://lists.llvm.org/pipermail/llvm-dev/2018-May/123252.html)). As explained in the accompanying slides, properly supporting linker scripts in LTO can avoid some correctness issues, but you'd imagine it might cause problems in this case.

Reply:
Thanks @asb. I'll check the link. I'll push another approach that uses a small assembly file along with a header with the expected size of the manifest. That should address my concerns regarding having the size in the linker script and also eliminate this macro. We can add the extern declaration to boot stages if needed in the future.

Reply:
Pushed a change that implements the idea above. PTAL.